### PR TITLE
nixos/acme: add alternative option to set credentials

### PR DESCRIPTION
**Pull request by datafoo**
This is to leverage systemd credentials for variables suffixed by _FILE.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/security/acme/default.nix [hunk 1]
```diff
@@ -566,6 +576,27 @@ let
...
+      credentials = mkOption {
+        type = types.attrsOf (types.str);
+        inherit (defaultAndText "credentials" {}) default defaultText;
+        description = lib.mdDoc ''
+          Environment variables to set for the cert's service for your selected
+          dnsProvider.
+          To find out what values you need to set, consult the documentation at
+          <https://go-acme.github.io/lego/dns/> for the corresponding dnsProvider.
+          This is an alternative to credentialsFile.
+          Variables suffixed by "_FILE" are passed securely with `LoadCredential=`.
```
Comment (SuperSandro2000) [inline nixos/modules/security/acme/default.nix:588]:
```suggestion
          Environment variables to set for the cert's service for your selected dnsProvider.
          To find out what values you need to set, consult the documentation at
          <https://go-acme.github.io/lego/dns/> for the corresponding dnsProvider.
          This is an alternative to credentialsFile.
          Variables suffixed by "_FILE" are passed securely with `LoadCredential=`.
```

**Comment by SuperSandro2000** [comment]
Why do we need an alternative? Is there something the other variant does not support? Can we unify them?
Having options makes the code more complex and harder to test

**Comment by m1cr0man** [comment]
I'm not sure what this solves for? The `environmentFile` is just that, and can be generated in your nix configs through pkgs.writeText if you plan to put environment variables in your configuration (not recommended), or can point to a file somewhere else on the file system for other use cases. Adding an option which explicitly enables a bad practice (AKA credentials in your nix store/configs if not using this _FILE prefix) seems like a bad idea to add, and for other use cases It's not clear what we gain over environmentFile.

**Comment by m1cr0man** [comment]
> (by leveraging systemd credentials, the file can be owned by root)

> Perhaps both options have their value then. And credentials could get better explanations and/or restriction on only variables suffixed by _FILE. What do you think?

In this case, there would be no difference between the existing facility (`environmentFile`) and you proposal. The source for `environmentFile` can already (and should) be owned as root with limited permissions. As such it still achieves the same thing. You can also add as many variables as you like to the file. Is there another difference I am overlooking?

File: nixos/modules/security/acme/default.nix (hunk 1 above)
Comment (emilazy) [reply nixos/modules/security/acme/default.nix:588]: No. This is fine. I don't see the point of this review.

**Comment by m1cr0man** [comment]
Ah I see :slightly_smiling_face:  There's a small misunderstanding here - environmentFile _can_ be owned by root:root with permissions as low as 0400. Just like credentialsFile, environmentFile is read by the systemd service manager before executing the service as the specified User=/Group=. As such it can be kept secure.

Just to double check, I made sure my secrets were owned by root:root and tested my own acme services on my server:
```
[root@web:~]# systemctl cat acme-m1cr0man.com.service | grep EnvironmentFile
EnvironmentFile=/run/secrets/acme_cloudflare_env

[root@web:~]# ls -al /run/secrets/acme_cloudflare_env 
-r-------- 1 root root 58 Jul 21 10:59 /run/secrets/acme_cloudflare_env
```

**Comment by m1cr0man** [comment]
Sorry no, we're not in agreement. To reiterate:

> By using systemd credentials, we expose a specific secret only to the 1 systemd unit that needs it, not the whole acme group. That is the difference!

This is already true for environmentFile, if its permissions are set correctly. It also inherently solves all three points you've highlighted in your last comment. No suffix check is required because all variables are in a secure file explicitly.

I do not want us to introduce another option which is loaded with a double duty, custom logic for handling those separate duties, and introduces an insecure path for configuring the certificate renewal credentials. At the very worst, a user can already set `systemd.services."acme-${name}".serviceConfig.Credentials` if they really want secrets in the store and don't want to use pkgs.writeText + environmentFile.

If what you want is an ability to set non-secret environment values like timeouts and such in your nix config, then the best solution I can think of is surfacing the fact that EnvironmentFile can be a list of paths or a single path, allowing you to use two files with a single cert - one with secrets and one with non secrets. This would simply be a case of changing the type on the environmentFile option.

**Comment by m1cr0man** [comment]
Oh I understand now. LoadCredential will fix the permissions securely. On that note - yes, I would prefer if all were adding here is support for the variables suffixed by _FILE. If you can drop conditional logic around that and drop support for non-_FILE creds then I think we're good to go 🙂

File: nixos/modules/security/acme/default.nix [hunk 2]
```diff
@@ -566,6 +572,22 @@ let
         example = "/var/src/secrets/example.org-route53-api-token";
       };
 
+      credentialFiles = mkOption {
```
Comment (m1cr0man) [inline nixos/modules/security/acme/default.nix:575]: This option name is way too similarly named to credentialsFile. I think the description needs to clear up what the difference is first - that this allows you to securely pass any secret file paths which lego must read at runtime by adding them to Systemd's LoadCredential, which will in turn set the permissions appropriately within its mount namespace.
Maybe `secretFiles` would work, as then a user could correlate that with the lego documentation when figuring out how to configure the module.

**Review by m1cr0man — CHANGES_REQUESTED** [review]
We need to figure out a better name for the option and a better description but technically it looks all good now :)

File: nixos/modules/security/acme/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply nixos/modules/security/acme/default.nix]: It is easier to read and we don't have a hard 80 character line limit

Comment (emilazy) [reply nixos/modules/security/acme/default.nix]: I'd really suggest focusing on more substantive review suggestions outside of cases of actively egregious/misleading formatting until there is any kind of agreed consensus on formatting rules (and automated tooling to apply and enforce them). There's no consistency about this in-tree and nitpicking about it with the present state of things just adds busywork for contributors without any real benefit.

**Review by m1cr0man — APPROVED** [review]
Sorry for the slow reply. The new option name looks good now :) happy to see this merged.

**Pull request merged**